### PR TITLE
refactor: FormDataProvider

### PR DESCRIPTION
**Pull request by Merkur39**
Now it has access to the id and label of the theme of the created file
This is for the purpose of future development

### CODE REVIEW

**Comment by Merkur39** [comment]
[Optimisation de la conversion en pdf (1/2j)](https://trello.com/c/cU94OXv9/763-optimisation-de-la-conversion-en-pdf-1-2j)

File: src/components/Contexts/FormDataProvider.jsx [hunk 1]
```diff
@@ -189,64 +32,22 @@ const FormDataProvider = ({ children }) => {
...
-            client
-          })
-        }
+        await createPdfAndSave({
+          formData,
+          qualification,
+          currentDefinition,
+          appFolderID,
+          client,
+          i18n: { t, f, scannerT }
```
Comment (Crash--) [inline src/components/Contexts/FormDataProvider.jsx:49]: Do we still really need `scannerT`?

File: src/helpers/buildFilename.spec.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import { buildFilename } from 'src/helpers/buildFilename'
+
+describe('buildFilename', () => {
+  it('should replace "/" by "|" in qualificationName', () => {
+    const paperName = buildFilename({
+      qualificationName: 'passport/test'
+    })
+
+    expect(paperName).toEqual('passport|test.pdf')
```
Comment (Crash--) [inline src/helpers/buildFilename.spec.js:9]: @taratatach Do we have a list of characters that may cause troubles for Desktop somewhere?

File: src/helpers/buildFilename.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/**
+ * @typedef {object} BuildPaperNameParam
+ * @property {string} qualificationName
+ * @property {string} [pageName]
+ * @property {string} [username]
+ * @property {string} [date]
```
Comment (Crash--) [inline src/helpers/buildFilename.js:6]: What is the format of the date?

File: src/helpers/createPdfAndSave.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    {
+      name,
+      contentType: 'application/pdf',
+      metadata,
+      dirId,
+      conflictStrategy: 'rename'
+    }
+  )
+
+  // The user is referenced as a contact on the new Paper
```
Comment (Crash--) [inline src/helpers/createPdfAndSave.js:71]: I don't think that the reference should be in this method. You should split it. 

The method signature will be easier. 

I don't even think the saveFile is necessary at all. We just need the method to add the reference.

File: src/helpers/createPdfAndSave.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  const references = contacts.map(contact => ({
+    _id: contact._id,
+    _type: CONTACTS_DOCTYPE
+  }))
+  await fileCollection.addReferencedBy(fileCreated, references)
```
Comment (Crash--) [inline src/helpers/createPdfAndSave.js:76]: no need to pass it as an argument. We already have the client you can build it from there

File: src/helpers/createPdfAndSave.js [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+  // eslint-disable-next-line no-unused-vars
+  const { multipage, ...newFileMetadata } = fileMetadata
+
+  return newFileMetadata
+}
+
+/**
+ * @param {SaveFileParam} param
+ * @returns {Promise<string>} Filename
```
Comment (Crash--) [inline src/helpers/createPdfAndSave.js:49]: This is not the filename that you return for now, but the id.

File: src/utils/resizeImage.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      canvas.width = scaledWidth
+      canvas.height = scaledHeight
+      canvas
+        .getContext('2d')
+        .drawImage(newImage, 0, 0, scaledWidth, scaledHeight)
+
+      resolve(canvas.toDataURL(fileType))
+    }
+  })
+}
```
Comment (Crash--) [inline src/utils/resizeImage.js:28]: nit: It's weird to have one method per file. I would have a created a utils/pdf.js with all the PDF related methods, and why not on utils/image.js to store image's related methods for instance

File: src/helpers/createPdfAndSave.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        opts: {
+          name: paperName,
+          metadata: newMetadata,
+          dirId: appFolderID
+        }
+      })
+      fileCreated.push({ fileId: fileIdCreated, themeLabel: label })
+    }
+
+    // If isn't multipage & not the last page, create new page of PDF
```
Comment (Ldoppea) [inline src/helpers/createPdfAndSave.js:148]: `PDFDocument.create()` is not about `page` creation but about `document` creation. So I'm not sure to understand the comment.

File: src/helpers/createPdfAndSave.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    if (!isMultiPage && idx !== data.length - 1) {
```
Comment (Ldoppea) [inline src/helpers/createPdfAndSave.js:149]: Is it possible to have a singlePage document that has an `idx !== data.length - 1` ?

File: src/helpers/createPdfAndSave.js [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+}) => {
+  const { t, f, scannerT } = i18n
+  const { data, metadata, contacts } = { ...formData }
+  const fileCollection = client.collection(FILES_DOCTYPE)
+  const isMultiPage = data.some(({ fileMetadata }) => fileMetadata.multipage)
+  const { featureDate, label } = currentDefinition
+  const date = metadata[featureDate] && f(metadata[featureDate], 'YYYY.MM.DD')
+  let fileCreated = []
+
+  // Created first page of pdf
```
Comment (Ldoppea) [inline src/helpers/createPdfAndSave.js:103]: This is not the first page of pdf. This is just a new empty pdf document.

Maybe we should avoid using the same term (page) for bot pdf document and `Mes papiers`'s form pages. Those are ambiguous.

Comment (Ldoppea) [reply src/helpers/createPdfAndSave.js:103]: Pdf pages are created inside of `addFileToPdf`

File: src/helpers/createPdfAndSave.js (hunk 8 above)
Comment (Ldoppea) [reply src/helpers/createPdfAndSave.js:148]: This seems to be related to the ambiguity between pdf page and form page. Isn't it?

File: src/helpers/buildFilename.spec.js (hunk 2 above)
Comment (taratatach) [reply src/helpers/buildFilename.spec.js:9]: Yes we do.

On Windows :
- forbidden chars are `<>:"/\|?*`
- forbidden last chars are `. `
- forbidden names are 'CON', 'PRN', 'AUX', 'NUL', 'COM1', 'COM2', 'COM3', 'COM4', 'COM5', 'COM6', 'COM7', 'COM8', 'COM9', 'LPT1', 'LPT2', 'LPT3', 'LPT4', 'LPT5', 'LPT6', 'LPT7', 'LPT8', 'LPT9'

On macOS and Linux : 
- forbidden chars are `/`

These can be found here: https://github.com/cozy-labs/cozy-desktop/blob/master/core/incompatibilities/platform.js#L110-L160

Comment (Crash--) [reply src/helpers/buildFilename.spec.js:9]: I knew it! Thank, I'll bookmark this file :D

File: src/components/Contexts/FormDataProvider.jsx (hunk 1 above)
Comment (Merkur39) [reply src/components/Contexts/FormDataProvider.jsx:49]: Yes, the name may be confusing, it now comes from `model.document` from `cozy-client`
https://github.com/cozy/mespapiers/blob/master/src/components/Contexts/ScannerI18nProvider.jsx#L3-L6

File: src/helpers/buildFilename.spec.js (hunk 2 above)
Comment (Merkur39) [reply src/helpers/buildFilename.spec.js:9]: Ok, I'll replace that.
Thx for the informations 👍

File: src/helpers/createPdfAndSave.js (hunk 9 above)
Comment (Merkur39) [reply src/helpers/createPdfAndSave.js:149]: No, it's not possible
`data` contains the file(s)
In the case of a single page, then the `idx` is always equal to `data.length - 1`

File: src/helpers/buildFilename.js (hunk 3 above)
Comment (Merkur39) [reply src/helpers/buildFilename.js:6]: I will rename this property and add a description to the jsDoc.
(This `date` is already formatted)

File: src/helpers/createPdfAndSave.js (hunk 8 above)
Comment (Merkur39) [reply src/helpers/createPdfAndSave.js]: Indeed, I renamed it

File: src/helpers/buildFilename.spec.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    ${{ qualificationName: 'passport' }}                                                                 | ${'passport.pdf'}
+    ${{ qualificationName: 'passport', pageName: 'front' }}                                              | ${'passport - front.pdf'}
+    ${{ qualificationName: 'passport', fullname: 'Bob' }}                                                | ${'passport - Bob.pdf'}
+    ${{ qualificationName: 'passport', formatedDate: '2022.01.01' }}                                     | ${'passport - 2022.01.01.pdf'}
+    ${{ qualificationName: 'passport', pageName: 'front', fullname: 'Bob' }}                             | ${'passport - front - Bob.pdf'}
+    ${{ qualificationName: 'passport', pageName: 'front', formatedDate: '2022.01.01' }}                  | ${'passport - front - 2022.01.01.pdf'}
+    ${{ qualificationName: 'passport', fullname: 'Bob', formatedDate: '2022.01.01' }}                    | ${'passport - Bob - 2022.01.01.pdf'}
+    ${{ qualificationName: 'passport', pageName: 'front', fullname: 'Bob', formatedDate: '2022.01.01' }} | ${'passport - front - Bob - 2022.01.01.pdf'}
```
Comment (Ldoppea) [inline src/helpers/buildFilename.spec.js:21]: image

File: src/helpers/buildFilename.js [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+/**
+ * Builded Paper name with qualification name & without use filename original
+ * @param {BuildPaperNameParam} opts
+ * @returns {string} Paper name with PDF extension
+ */
+export const buildFilename = opts => {
+  /*
+    Calling the stack's file creation method would trigger a ``status: "422", title: "Invalid Parameter"error if filename contains/`.
```
Comment (Ldoppea) [inline src/helpers/buildFilename.js:16]: Seems like there are some wrongly placed backquotes
```suggestion
    Calling the stack's file creation method would trigger a `status: "422", title: "Invalid Parameter"` error if filename contains `/`.
```

**Review by Ldoppea — APPROVED** [review]
Just a small typo to fix (see comment), then everything looks great to me 👍

**Comment by Merkur39** [comment]
Just added this comment
https://github.com/cozy/mespapiers/pull/137/files#diff-e6049e9ea657c5d873f58abb455718efa78a0dddb4054ac64488b70c34704b01R80-R81

**Pull request merged**